src/Data/Fin/Properties.agda

```diff
@@ -570,6 +587,76 @@ combine-remQuot {suc n} k i with splitAt k i | P.inspect (splitAt k) i
   i                                        ∎
   where open ≡-Reasoning
 
+toℕ-combine : ∀ {n k} (x : Fin n) (y : Fin k) → toℕ (combine x y) ≡ k ℕ.* toℕ x ℕ.+ toℕ y
+toℕ-combine {n = suc n} {k} x@0F y = begin
+  toℕ (combine x y)          ≡⟨⟩
+  toℕ (y ↑ˡ (n ℕ.* k))       ≡⟨ toℕ-↑ˡ y (n ℕ.* k) ⟩
+  toℕ y                      ≡⟨⟩
+  0 ℕ.+ toℕ y                ≡˘⟨ cong (ℕ._+ toℕ y) (ℕₚ.*-zeroʳ k) ⟩
+  k ℕ.* toℕ x ℕ.+ toℕ y      ∎
+  where open ≡-Reasoning
+toℕ-combine {n = suc n} {k} (suc x) y = begin
+  toℕ (combine (suc x) y)        ≡⟨⟩
+  toℕ (k ↑ʳ combine x y)         ≡⟨ toℕ-↑ʳ k (combine x y) ⟩
+  k ℕ.+ toℕ (combine x y)        ≡⟨ cong (k ℕ.+_) (toℕ-combine x y) ⟩
+  k ℕ.+ (k ℕ.* toℕ x ℕ.+ toℕ y)  ≡⟨ solve 3 (λ k x y → k :+ (k :* x :+ y) := k :* (con 1 :+ x) :+ y) refl k (toℕ x) (toℕ y) ⟩
+  k ℕ.* toℕ (suc x) ℕ.+ toℕ y    ∎
+  where
+    open ≡-Reasoning
+    open import Data.Nat.Solver
+    open +-*-Solver
+
+combine-injectiveˡ : ∀ {n k} (x y : Fin n) (z : Fin k) → combine x z ≡ combine y z → x ≡ y
+combine-injectiveˡ {n} {k@(suc _)} x y z combine[x,z]≡combine[y,z] = toℕ-injective (ℕₚ.*-cancelˡ-≡ k (ℕₚ.+-cancelʳ-≡ (k ℕ.* toℕ x) (k ℕ.* toℕ y) (begin
+  k ℕ.* toℕ x ℕ.+ toℕ z      ≡˘⟨ toℕ-combine x z ⟩
+  toℕ (combine x z)          ≡⟨ cong toℕ combine[x,z]≡combine[y,z] ⟩
+  toℕ (combine y z)          ≡⟨ toℕ-combine y z ⟩
+  k ℕ.* toℕ y ℕ.+ toℕ z      ∎)))
+  where open ≡-Reasoning
+
+combine-injectiveʳ : ∀ {n k} (x : Fin n) (y z : Fin k) → combine x y ≡ combine x z → y ≡ z
+combine-injectiveʳ {n} {k} x y z combine[x,z]≡combine[y,z] = toℕ-injective (ℕₚ.+-cancelˡ-≡ (k ℕ.* toℕ x) (begin
+  k ℕ.* toℕ x ℕ.+ toℕ y ≡˘⟨ toℕ-combine x y ⟩
+  toℕ (combine x y)     ≡⟨ cong toℕ combine[x,z]≡combine[y,z] ⟩
+  toℕ (combine x z)     ≡⟨ toℕ-combine x z ⟩
+  k ℕ.* toℕ x ℕ.+ toℕ z ∎))
+  where open ≡-Reasoning
+
+combine-injective : ∀ {n k} (x : Fin n) (y : Fin k) (w : Fin n) (z : Fin k) → combine x y ≡ combine w z → x ≡ w × y ≡ z
+combine-injective x y w z combine[x,y]≡combine[w,z] = lemma₂ x y w z combine[x,y]≡combine[w,z] , lemma₃ x y w z combine[x,y]≡combine[w,z] where
```

Comment:
The where should definitely go on a new line, I completely missed it all the way over there off my screen. Was trying to work out why the indentation increased :smile: